### PR TITLE
feat(cdn): add POST endpoint

### PR DESCRIPTION
**Pull request by JacopoDaeli**
Still draft, but ready for initial feedback gathering. 

image

### CODE REVIEW

File: lib/plugins/cdn.js [hunk 1]
```diff
@@ -1,10 +1,128 @@
...
+          stream
+            .pipe(tar.t())
+            .on('entry', async (entry) => {
+              const {
+                header: { path: filename }
+              } = entry;
+              fastify.log.info(`Processing file ${filename}`);
+              const buffer = await fastify.streamToBuffer(entry);
+              if (filename === '_metadata.json') {
+                metadata = JSON.parse(buffer.toString('utf8'));
```
Comment (decompil3d) [inline lib/plugins/cdn.js:68]: Should we be more defensive to bad JSON here?

File: lib/plugins/cdn.js [hunk 2]
```diff
@@ -1,10 +1,128 @@
...
+      }
+    );
+
+    fastify.decorate(
+      'uploadFileToStorage',
+      /**
+       * Upload file to S3 storage service.
+       * @param {Object} opts Method parameters
+       * @param {string} opts.filepath File path
+       * @param {number} [opts.exipration] Expiration timestamp
```
Comment (decompil3d) [inline lib/plugins/cdn.js:89]:
```suggestion
       * @param {number} [opts.expiration] Expiration timestamp
```

File: lib/plugins/cdn.js [hunk 3]
```diff
@@ -1,10 +1,128 @@
...
+       * @param {Buffer} opts.data File data
+       * @returns {Promise<any>} Operation result
```
Comment (decompil3d) [inline lib/plugins/cdn.js:91]: Can we use the type that `aws-sdk` exports for the `PutObject` call?

File: lib/plugins/cdn.js [hunk 4]
```diff
@@ -1,10 +1,128 @@
...
+      }
+    );
+
+    fastify.decorate(
+      'uploadAssetFilesToStorage',
+      /**
+       * Upload Warehouse assets to storage.
+       * @param {WrhsAssetFile[]} files - Warehouse asset files
+       * @param {number} [expiration] - Assets expiration timestamp
+       * @returns {Promise<any>} Operation result
```
Comment (decompil3d) [inline lib/plugins/cdn.js:111]: Same -- can we use a proper type here?

**Review by decompil3d — CHANGES_REQUESTED** [review]

**Comment by JacopoDaeli** [comment]
Probably in enough good shape now.

File: lib/plugins/cdn.js [hunk 5]
```diff
@@ -1,10 +1,135 @@
...
+        await new Promise((resolve, reject) => {
+          stream
+            .pipe(tar.t())
+            .on('entry', async (entry) => {
+              const {
+                header: { path: filename }
+              } = entry;
+              fastify.log.info(`Processing file ${filename}`);
+              const buffer = await fastify.streamToBuffer(entry);
+              if (filename === '_metadata.json') {
```
[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/cdn.js:68]: No it's not a convention to mark `private/hidden`. It's just a convention for the metadata file.

File: lib/plugins/cdn.js [hunk 6]
```diff
@@ -1,10 +1,135 @@
...
+                const unsafeJson = buffer.toString('utf8');
+                fastify.log.debug(unsafeJson);
+                try {
+                  metadata = JSON.parse(unsafeJson);
+                } catch (err) {
+                  fastify.log.warning(err);
+                  reject(err);
```
[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/cdn.js:76]: Stream should not be dropped since it's the http request itself. I'll remove the `on('entry')` listener to be extra safe.

**Comment by JacopoDaeli** [comment]
@decompil3d looks good to you now?

**Review by decompil3d — APPROVED** [review]

**Pull request merged**